// emitter-array/build.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);throw new Error("Cannot find module '"+o+"'")}var f=n[o]={exports:{}};t[o][0].call(f.exports,function(e){var n=t[o][1][e];return s(n?n:e)},f,f.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
var Emitter = require('emitter-component');

/**
 * Proxied array mutators methods
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */
var mutators = {
  pop: function() {

    var ret = Array.prototype.pop.apply(this);

    this.emit('popped', ret);
    this.emit('change');

    return ret;
  },
  push: function() {

    var ret = Array.prototype.push.apply(this, arguments);

    this.emit('popped', ret);
    this.emit('change');

    return ret;
  },
  reverse: function() {

    var ret = Array.prototype.reverse.apply(this);

    this.emit('reversed', ret);
    this.emit('change');

    return ret;
  },
  shift: function() {

    var ret = Array.prototype.shift.apply(this);

    this.emit('shifted', ret);
    this.emit('change');

    return ret;
  },
  sort: function() {

    var ret = Array.prototype.sort.apply(this);

    this.emit('sorted', ret);
    this.emit('change');

    return ret;
  },
  unshift: function() {

    var ret = Array.prototype.unshift.apply(this, arguments);

    this.emit('unshifted', ret);
    this.emit('change');

    return ret;
  },
  splice: function() {

    if (!arguments.length) {
      return;
    }

    var newItems = Array.prototype.slice.call(arguments, 2);

    var ret = Array.prototype.splice.apply(this, arguments);

    this.emit('spliced', ret);
    this.emit('change');

    return ret;
  }
};

window.emitterArray = module.exports = function() {

  /**
   * Construct an Array from the passed arguments
   */
  var arrCtorArgs = arguments;
  var arr = Array.apply(null, arrCtorArgs);

  /**
   * Mixin Emitter to the Array instance
   */
  Emitter(arr);

  /**
   * Proxy all Array.prototype mutator methods on this array instance
   */
  arr.pop = arr.pop && mutators.pop.bind(arr);
  arr.push = arr.push && mutators.push.bind(arr);
  arr.reverse = arr.reverse && mutators.reverse.bind(arr);
  arr.shift = arr.shift && mutators.shift.bind(arr);
  arr.sort = arr.sort && mutators.sort.bind(arr);
  arr.splice = arr.splice && mutators.splice.bind(arr);


  return arr;
};
},{"emitter-component":2}],2:[function(require,module,exports){

/**
 * Expose `Emitter`.
 */

module.exports = Emitter;

/**
 * Initialize a new `Emitter`.
 *
 * @api public
 */

function Emitter(obj) {
  if (obj) return mixin(obj);
};

/**
 * Mixin the emitter properties.
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */

function mixin(obj) {
  for (var key in Emitter.prototype) {
    obj[key] = Emitter.prototype[key];
  }
  return obj;
}

/**
 * Listen on the given `event` with `fn`.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.on =
Emitter.prototype.addEventListener = function(event, fn){
  this._callbacks = this._callbacks || {};
  (this._callbacks[event] = this._callbacks[event] || [])
    .push(fn);
  return this;
};

/**
 * Adds an `event` listener that will be invoked a single
 * time then automatically removed.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.once = function(event, fn){
  var self = this;
  this._callbacks = this._callbacks || {};

  function on() {
    self.off(event, on);
    fn.apply(this, arguments);
  }

  on.fn = fn;
  this.on(event, on);
  return this;
};

/**
 * Remove the given callback for `event` or all
 * registered callbacks.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.off =
Emitter.prototype.removeListener =
Emitter.prototype.removeAllListeners =
Emitter.prototype.removeEventListener = function(event, fn){
  this._callbacks = this._callbacks || {};

  // all
  if (0 == arguments.length) {
    this._callbacks = {};
    return this;
  }

  // specific event
  var callbacks = this._callbacks[event];
  if (!callbacks) return this;

  // remove all handlers
  if (1 == arguments.length) {
    delete this._callbacks[event];
    return this;
  }

  // remove specific handler
  var cb;
  for (var i = 0; i < callbacks.length; i++) {
    cb = callbacks[i];
    if (cb === fn || cb.fn === fn) {
      callbacks.splice(i, 1);
      break;
    }
  }
  return this;
};

/**
 * Emit `event` with the given args.
 *
 * @param {String} event
 * @param {Mixed} ...
 * @return {Emitter}
 */

Emitter.prototype.emit = function(event){
  this._callbacks = this._callbacks || {};
  var args = [].slice.call(arguments, 1)
    , callbacks = this._callbacks[event];

  if (callbacks) {
    callbacks = callbacks.slice(0);
    for (var i = 0, len = callbacks.length; i < len; ++i) {
      callbacks[i].apply(this, args);
    }
  }

  return this;
};

/**
 * Return array of callbacks for `event`.
 *
 * @param {String} event
 * @return {Array}
 * @api public
 */

Emitter.prototype.listeners = function(event){
  this._callbacks = this._callbacks || {};
  return this._callbacks[event] || [];
};

/**
 * Check if this emitter has `event` handlers.
 *
 * @param {String} event
 * @return {Boolean}
 * @api public
 */

Emitter.prototype.hasListeners = function(event){
  return !! this.listeners(event).length;
};

},{}]},{},[1])
